### PR TITLE
Fix division calculation in `dask.dataframe.io.from_dask_array`

### PR DESCRIPTION
**Pull request by jjyap**
The divisions calculated in `dask.dataframe.io.from_dask_array` are incorrect if you pass in an array with empty chunks at the end.

E.g If you pass a dask array with `chunks = (1, 2, 3, 0)` to `from_dask_array` the divisions calculated are `(0, 1, 3, 6, 5)`.  This PR correctly handles the special case of empty chunks at the end and calculates the divisions as `[0, 1, 3, 5, 5]`.

I also added some tests to check empty chunks at the beginning and middle (these worked fined pre-PR).

- [ ] Closes #xxxx
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by jyap-abc** [comment]
😬 looks like this touches a lot more than I had hoped for.  Any thoughts about the failing tests?

**Comment by jjyap** [comment]
Any thoughts about the failing tests?

**Comment by jjyap** [comment]
Hi @jrbourbeau, would you mind taking a look at this. This solves an annoying issue I'm running into.

**Comment by pavithraes** [comment]
@jjyap Good catch, and thank you for this PR!

I looked at this with @ncclementi, and we think your implementation is clever! However, we might be calculating things more than once (e.g., with `cumsum`) and doing expensive-ish conversions (e.g., list to `pd.Series`).

Maybe we can keep what we had earlier, but just make it work for this edge case of having empty chunks at the end?

Something like:

```python
else:
        # The shape of the incoming array is known and we don't have an explicit index.
        # Create a mapping of chunk number in the incoming array to
        # (start row, stop row) tuples. These tuples will be used to create a sequential
        # RangeIndex later on that is continuous over the whole DataFrame.
        n_elements = sum(x.chunks[0])
        divisions = [0]
        stop = 0
        index_mapping = {}
        for i, increment in enumerate(x.chunks[0]):
            stop += increment
            index_mapping[(i,)] = (divisions[i], stop)

            # last division corrected, even if there are empty chunk(s) at the end
            if stop == n_elements:
                stop -= 1

            divisions.append(stop)

        arrays_and_indices.extend([BlockwiseDepDict(mapping=index_mapping), "i"])
```

We tested this locally, and it seems to work. What do you think?

File: dask/dataframe/io/tests/test_io.py [hunk 1]
```diff
@@ -553,6 +553,29 @@ def test_from_dask_array_unknown_chunks():
     assert_eq(df, pd.DataFrame(dx.compute()), check_index=False)
 
 
+def test_from_dask_array_empty_chunks():
+    # Empty chunks at the end
+    chunksizes = (1, 2, 3, 0)
+    monotonic_index = da.from_array(np.arange(sum(chunksizes)), chunks=chunksizes)
```
Comment (pavithraes) [inline dask/dataframe/io/tests/test_io.py:559]:
```suggestion
    monotonic_index = da.from_array(np.arange(6), chunks=chunksizes)
```

Since it's alwasy 6 elements throughout this test. :)

File: dask/dataframe/io/tests/test_io.py [hunk 2]
```diff
@@ -553,6 +553,29 @@ def test_from_dask_array_unknown_chunks():
...
+
+    df = dd.from_dask_array(monotonic_index)
+    assert df.divisions == (0, 1, 3, 5, 5)
+
+    # Empty chunks at the beginning
+    chunksizes = (0, 1, 2, 3)
```
Comment (pavithraes) [inline dask/dataframe/io/tests/test_io.py:565]: This and the next assert are testing similar things. Maybe can use parameterization here instead?

In this case, it will look something like:

```python
@pytest.mark.parametrize(
    "chunksizes, expected_divisions",
    [
        pytest.param(
            (1, 2, 3, 0), (0, 1, 3, 5, 5)
        ),
        pytest.param(
            (0, 1, 2, 3), (0, 0, 1, 3, 5)
        ),
        pytest.param(
            (1, 0, 2, 3), (0, 1, 1, 3, 5)
        ),
    ]
)
def test_from_dask_array_empty_chunks(chunksizes, expected_divisions):
    # Empty chunks at the end
    monotonic_index = da.from_array(np.arange(6), chunks=chunksizes)

    df = dd.from_dask_array(monotonic_index)
    assert df.divisions == expected_divisions
```

**Review by ncclementi — COMMENTED** [review]
I reviewed this with @pavithraes, the PR looks good we just left a couple of comments and suggestions.

**Comment by jjyap** [comment]
@pavithraes @ncclementi thank you both!

Avoiding a `pd.Series` conversion makes sense to me. I like your suggestions and included them as is.

**Comment by jyap-abc** [comment]
> @jjyap Good catch, and thank you for this PR!
> 
> I looked at this with @ncclementi, and we think your implementation is clever! However, we might be calculating things more than once (e.g., with `cumsum`) and doing expensive-ish conversions (e.g., list to `pd.Series`).
> 
> Maybe we can keep what we had earlier, but just make it work for this edge case of having empty chunks at the end?
> 
> Something like:
> 
> ```python
> else:
>         # The shape of the incoming array is known and we don't have an explicit index.
>         # Create a mapping of chunk number in the incoming array to
>         # (start row, stop row) tuples. These tuples will be used to create a sequential
>         # RangeIndex later on that is continuous over the whole DataFrame.
>         n_elements = sum(x.chunks[0])
>         divisions = [0]
>         stop = 0
>         index_mapping = {}
>         for i, increment in enumerate(x.chunks[0]):
>             stop += increment
>             index_mapping[(i,)] = (divisions[i], stop)
> 
>             # last division corrected, even if there are empty chunk(s) at the end
>             if stop == n_elements:
>                 stop -= 1
> 
>             divisions.append(stop)
> 
>         arrays_and_indices.extend([BlockwiseDepDict(mapping=index_mapping), "i"])
> ```
> 
> We tested this locally, and it seems to work. What do you think?

Wait I think this will compute something different than I proposed.  We shouldn't store the value of stop after we decrement it.

**Comment by jyap-abc** [comment]
Oh nevermind, this actually works my bad!  Stop will only get decremented once.

**Comment by pavithraes** [comment]
@jjyap Awesome! Thanks for working on this!
